### PR TITLE
Add bufq fuzzing harness

### PR DESCRIPTION
**Pull request by elopez**
This PR refactors the existing TLV handling code to make it a bit more reusable. It then adds a new standalone harness that fuzzes the bufq buffer handling.

### CODE REVIEW

**Comment by cmeister2** [comment]
I'll dig in a bit more in depth later, but my top-level comment is that you don't necessarily need to keep to the TLV format for fuzzer test cases if it doesn't make sense. Does https://github.com/google/fuzzing/blob/master/docs/split-inputs.md#fuzzed-data-provider make sense for this fuzzer?

File: curl_fuzzer.h [hunk 1]
```diff
@@ -23,61 +23,11 @@
...
-#define TLV_TYPE_USERPWD			44
-#define TLV_TYPE_USERAGENT			45
-#define TLV_TYPE_NETRC				46
-#define TLV_TYPE_SSH_HOST_PUBLIC_KEY_SHA256	47
-#define TLV_TYPE_POST				48
-#define TLV_TYPE_WS_OPTIONS			49
-#define TLV_TYPE_CONNECT_ONLY			50
-#define TLV_TYPE_HSTS				51
-#define TLV_TYPE_HTTPPOSTBODY			52
+#ifdef TLV_ENUM_BUFQ
```
Comment (cmeister2) [inline curl_fuzzer.h:26]: This smells bad - if we're having to define two flavours of TLV header file it probably means that the TLV framework isn't the right fit for the fuzzer.

File: mainline.sh [hunk 2]
```diff
@@ -31,10 +31,10 @@ shift $((OPTIND-1))
 export CC=clang
 export CXX=clang++
 FUZZ_FLAG="-DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION"
-export CFLAGS="-fsanitize=address"
-export CXXFLAGS="-fsanitize=address -stdlib=libstdc++ $FUZZ_FLAG"
+export CFLAGS="-fsanitize=address,fuzzer-no-link"
```
Comment (cmeister2) [inline mainline.sh:34]: Why are these required?

File: curl_fuzzer_tlv_bufq.cc [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+    case TLV_TYPE_NO_SPARE:
+      FCHECK(tlv->length == 0);
+      FCHECK(fuzz->no_spare == 0);
+      fuzz->no_spare = 1;
+      break;
+
+    case TLV_TYPE_RESET:
+      tmp_op_type = OP_TYPE_RESET;
+      goto ADD_OP_UNSIZED;
```
Comment (cmeister2) [inline curl_fuzzer_tlv_bufq.cc:76]: Not convinced about using `goto`s here.

I think switching out this whole mechanism for the FuzzedDataProvider makes more sense given the classes of operations you're trying to do here. With that we could in a loop do:

- get a byte for the operation type
- get extra bytes if required for the operation
- operate directly on the buffer (no creating operation lists)
- continue until FuzzedDataProvider is empty.

It'd also be a smidge faster because you're not having to precalculate the operations list (and allocating memory for it).

**Review by cmeister2 — COMMENTED** [review]
I see the vision here and in general I'm a lot happier with this PR. I'm dubious as to whether a TLV mechanism here is "correct", though.

File: mainline.sh (hunk 2 above)
Comment (elopez) [reply mainline.sh:34]: `fuzzer-no-link` as added as the harnesses weren't finding any coverage when running. After adding this, the warning went away.

```
WARNING: no interesting inputs were found so far. Is the code instrumented for coverage?
This may also happen if the target rejected all inputs we tried so far
```

As for linking stdc++, OpenSSL fails to build on my Ubuntu system without linking it:

```
${LDCMD:-clang-17} -pthread -m64 -fno-omit-frame-pointer -g -Wa,--noexecstack -Qunused-arguments -fsanitize=address,fuzzer-no-link -fsanitize=address,fuzzer-no-link -fno-sanitize=alignment -L../../usr/lib -L.   \
        -o fuzz/asn1 \
        fuzz/asn1-bin-asn1.o fuzz/asn1-bin-driver.o \
        fuzz/asn1-bin-fuzz_rand.o \
        -lFuzzingEngine -lssl -lcrypto -ldl -pthread 
/usr/bin/ld: ../../usr/lib/libFuzzingEngine.a(FuzzerIO.cpp.o): in function `fuzzer::FileToVector(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned long, bool)':
(.text._ZN6fuzzer12FileToVectorERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEmb+0x2b): undefined reference to `std::basic_ifstream<char, std::char_traits<char> >::basic_ifstream(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::_Ios_Openmode)'
// many many more lines (...)
```

I'm using clang 17 and the following for the fuzzing engine:
```
lrwxrwxrwx 1 root root 67 Dec 13 15:58 /usr/lib/libFuzzingEngine.a -> /usr/lib/llvm-17/lib/clang/17/lib/linux/libclang_rt.fuzzer-x86_64.a
```

File: curl_fuzzer_tlv_bufq.cc (hunk 3 above)
Comment (elopez) [reply curl_fuzzer_tlv_bufq.cc]: Hi @cmeister2, I reverted the TLV changes and reworked this with FuzzedDataProvider, let me know if that looks any better 👍  execs/s as reported by libfuzzer are significantly reduced, but I assume the TLV approach had many more 'rejected' executions that are shorter.

File: mainline.sh (hunk 2 above)
Comment (cmeister2) [reply mainline.sh:34]: I'm just a bit surprised it's needed. Have you tried building the fuzzers with the oss-fuzz images or just locally?

File: curl_fuzzer_tlv_bufq.cc (hunk 3 above)
Comment (cmeister2) [reply curl_fuzzer_tlv_bufq.cc]: This looks awesome, thanks for implementing it! I don't know the internal mechanics of FuzzedDataProvider but I hope it's optimised for this kind of usage.

File: mainline.sh (hunk 2 above)
Comment (elopez) [reply mainline.sh:34]: building through the oss-fuzz scripts seemed to be fine when I tried, but that goes through a different script (`ossfuzz.sh`). I haven't compared the environments to see what the difference is; it's possible it might be related to the compiler or fuzzing engine versions I'm using.

Comment (cmeister2) [reply mainline.sh:34]: Ok. Happy with this then. If you can fix up the merge conflicts I'll get this merged in.
